File: manual-filters/filters-py/filter.py
''' Module: filter

    Classes contained: Filter

    Created by S.T. Castle on 2015-01-16

    Create filters for image convolutions.
'''

import string
import math

class Filter:
    '''A box, tent, or gaussian filter.'''

    # Constructor.
    # params: filter_type: string with value BOX, TENT, or GAUSSIAN.
    #         filter_r: integer radius of the filter.
    # throws: KeyError
    def __init__(self, filter_type, filter_r):
        self.type = string.upper(filter_type)  # Filter type.
        self.r = filter_r                      # Filter radius.
        self.w = 2*self.r + 1;                 # Filter width.
        self.f = []                            # Filter values.
        
        # Stores the available filter types.
        types = {
            'BOX'      : self._make_box,
            'TENT'     : self._make_tent,
            'GAUSSIAN' : self._make_gaussian,
        }

        # Make the filter.
        try:
            types[self.type]()
        except KeyError:
            print 'ERROR: Filter type \'' +filter_type+ '\' is not supported.'

    # Methods to make the filter of the specified type.

    def _make_box(self):
        val = 1.0 / self.w  # Normalized, uniform values.
        # Put values into filter with width w.
        for i in range(self.w):
            self.f.append(val)

    def _make_tent(self):
        sum = 0.0  # Track the sum of filter values for normalization.

        # Compute the value at each position in the filter.
        for i in range(self.w):
            x = abs(i-self.r)  # Displacement from center of filter.
            val = (1.0-(x*1.0/self.r))/self.r
            self.f.append(val)
            sum += val

        # Normalize values in the filter.
        for i in range(self.w):
            self.f[i] = self.f[i] / sum

    def _make_gaussian(self):
        sum = 0.0  # Track the sum of filter values for normalization.
        sigma = (self.r - 1.0) / 3.0; # Standard dev for Gaussian distro.

        # Compute filter values.
        for i in range(self.w):
            # Calculate vale for filter at position i-radius.
            # This is the displacement from the center pixel (at index self.r).
            val = self._gauss(i-self.r, sigma)
            self.f.append(val)
            sum += val

        # Normalize values in the filter.
        for i in range(self.w):
            self.f[i] = self.f[i] / sum

    # Gaussian function.
    def _gauss(self, x, sig):
        return ((1.0/(sig*math.sqrt(2*math.pi)))*math.exp(-x*x/(2*sig*sig)))

    # Accessor Methods

    # Return the filter type.
    def get_type(self):
        return self.type

    # Return the filter radius.
    def get_radius(self):
        return self.r

    # Return the filter list.
    def get_filter(self):
        return self.f

File: manual-filters/filters-cpp/Filter.cpp
//
//  Filter.h
//  ppmReaderWriter
//
//  Created by Castle, Sam on 10/8/14.
//  Davidson College CSC 361 Fall 2014 - HW 4
//  Copyright (c) 2014 Davidson College. All rights reserved.
//
//  Class creates a filter for image convolutions.
//  header file

#include <cmath>
#include "Filter.h"


// constructs a 1D filter object of the specified filter_type (box, tent, or
// gaussian) with radius filter_r. So, total filter width is (2*filter_r + 1).
Filter::Filter(filter_name filter_type, int filter_r) {

    f_radius = filter_r;
    f_width = 2*f_radius + 1;
    f_array = new float[f_width];
    
    switch (filter_type) {
        case BOX: make_box(); break;
        case TENT: make_tent(); break;
        case GAUSSIAN: make_gaussian(); break;
    }
    
}

Filter::~Filter() {
    delete [] f_array;
}

float* Filter::getFilter() {
    return f_array;
}

int Filter::getRadius() {
    return f_radius;
}

int Filter::getWidth() {
    return f_width;
}

// pre: f_array is defined and has size f_width
void Filter::make_box() {

    float val = 1.0 / f_width; // normalized value at each point
    for (int i=0; i<f_width; i++)
        f_array[i] = val;
    
}

// pre: f_array is defined and has size f_width
void Filter::make_tent() {

    float sum = 0.0; // track the sum of filter values for normalization
    
    // fill f_array
    for (int i=0; i<f_width; i++) {
        // displacement from center pixel, which has index f_radius
        float x = i - f_radius;
        // compute the tent filter value for the current position
        float val = ( 1.0 - abs(x*1.0/f_radius) ) / f_radius;
        // place val in array and increase sum
        f_array[i] = val;
        sum += val;
    }
    
    // normalize values in f_array
    for (int i=0; i<f_width; i++)
        f_array[i] = f_array[i] / sum;

}

// pre: f_array is defined and has size f_width
void Filter::make_gaussian() {

    float sum = 0.0; // track the sum of filter values for normalization
    float sigma = (f_radius - 1.0) / 3.0; // standard dev for gaussian distro
    
    // fill f_array
    for (int i=0; i<f_width; i++) {
        // calculate value for filter at position i-f_radius, which equals the
        // displacement from the center pixel located at index f_radius
        float val = gauss(i-f_radius, sigma);
        // place val in array and increase sum
        f_array[i] = val;
        sum += val;
    }
    
    // normalize values in f_array
    for (int i=0; i<f_width; i++)
        f_array[i] = f_array[i] / sum;
    
}

// returns the value at position x of a gaussian distribution centered at zero
// and with standard deviation sigma
float Filter::gauss(float x, float sigma) {
    return ( (1.0 / (sigma * sqrt(2*M_PI)) ) * exp(-x*x / (2*sigma*sigma)) );
}